src/chain.h

```diff
@@ -414,16 +414,9 @@ class CDiskBlockIndex : public CBlockIndex
         return block.GetHash();
     }
 
-
-    std::string ToString() const
-    {
-        std::string str = "CDiskBlockIndex(";
-        str += CBlockIndex::ToString();
-        str += strprintf("\n                hashBlock=%s, hashPrev=%s)",
-            GetBlockHash().ToString(),
-            hashPrev.ToString());
-        return str;
-    }
+    // To be called only by the base class.
```

Comment:
I am pretty sure that constructing a `CDiskBlockIndex`, then calling the base class's method on the constructed object will result in dereferencing a nullptr (UB)

Reply:
I don't get it. Do you mean this?

```cpp
class B
{
public:
    void f() { std::cout << "B::f()\n"; }
};

class D : public B
{
public:
    void f() = delete;
};

int main(int, char**)
{
    D d;
    static_cast<B*>(&d)->f();
    return 0;
}
```

(it prints "B::f()")

Reply:
Running mainnet with the following patch:
```diff
diff --git a/src/txdb.cpp b/src/txdb.cpp
index bad3bb80a9..5078ddb6bf 100644
--- a/src/txdb.cpp
+++ b/src/txdb.cpp
@@ -310,7 +310,9 @@ bool CBlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams,
             CDiskBlockIndex diskindex;
             if (pcursor->GetValue(diskindex)) {
                 // Construct block index object
-                CBlockIndex* pindexNew = insertBlockIndex(diskindex.ConstructBlockHash());
+                CBlockIndex* pindexNew{insertBlockIndex(diskindex.ConstructBlockHash())};
+                assert(pindexNew);
+                LogPrintf("CBI 1: %s\n", pindexNew->ToString());
                 pindexNew->pprev          = insertBlockIndex(diskindex.hashPrev);
                 pindexNew->nHeight        = diskindex.nHeight;
                 pindexNew->nFile          = diskindex.nFile;
@@ -323,6 +325,7 @@ bool CBlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams,
                 pindexNew->nNonce         = diskindex.nNonce;
                 pindexNew->nStatus        = diskindex.nStatus;
                 pindexNew->nTx            = diskindex.nTx;
+                LogPrintf("CBI 2: %s\n", pindexNew->ToString());
 
                 if (!CheckProofOfWork(pindexNew->GetBlockHash(), pindexNew->nBits, consensusParams)) {
                     return error("%s: CheckProofOfWork failed: %s", __func__, pindexNew->ToString());
```
prints (many) valid-looking objects.
```
2022-07-15T16:34:08Z CBI 1: CBlockIndex(pprev=0, nHeight=0, merkle=0000000000000000000000000000000000000000000000000000000000000000, hashBlock=000000000000000000597281aa1ec7e99f0fff6eeda0db8707f77709951ba61a)
2022-07-15T16:34:08Z CBI 2: CBlockIndex(pprev=0x55d02f11c778, nHeight=478600, merkle=40e759d1ea04b507003f1177a66c39ec2783eb464f7d867c3d7f401fb168f2aa, hashBlock=000000000000000000597281aa1ec7e99f0fff6eeda0db8707f77709951ba61a)

2022-07-15T16:34:08Z CBI 1: CBlockIndex(pprev=0, nHeight=0, merkle=0000000000000000000000000000000000000000000000000000000000000000, hashBlock=00000000000000000506e18c023143c5c3421399d3582be8aa4b263b8f69a61a)
2022-07-15T16:34:08Z CBI 2: CBlockIndex(pprev=0x55d02f11c918, nHeight=427311, merkle=29af1206ec66690b1a0e51c2b07e6ead88056e7c11746c4d448f47e161cbbe39, hashBlock=00000000000000000506e18c023143c5c3421399d3582be8aa4b263b8f69a61a)
```
and with this patch calling the base function that is deleted in the derived class:
```diff
diff --git a/src/txdb.cpp b/src/txdb.cpp
index bad3bb80a9..c048c2d92a 100644
--- a/src/txdb.cpp
+++ b/src/txdb.cpp
@@ -310,7 +310,7 @@ bool CBlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams,
             CDiskBlockIndex diskindex;
             if (pcursor->GetValue(diskindex)) {
                 // Construct block index object
-                CBlockIndex* pindexNew = insertBlockIndex(diskindex.ConstructBlockHash());
+                CBlockIndex* pindexNew = insertBlockIndex(diskindex.GetBlockHash());
                 pindexNew->pprev          = insertBlockIndex(diskindex.hashPrev);
                 pindexNew->nHeight        = diskindex.nHeight;
```
I see the following build error:
```
```
txdb.cpp:313:69: error: attempt to use a deleted function
                CBlockIndex* pindexNew = insertBlockIndex(diskindex.GetBlockHash());
                                                                    ^
./chain.h:418:13: note: 'GetBlockHash' has been explicitly marked deleted here
    uint256 GetBlockHash() = delete;
            ^
1 error generated.
```
Let me know if I'm missing something.

Reply:
I mean that calling the base class method on a disk index object is UB:

```diff
diff --git a/src/txdb.cpp b/src/txdb.cpp
index c048c2d92a..de1384fb61 100644
--- a/src/txdb.cpp
+++ b/src/txdb.cpp
@@ -309,6 +309,7 @@ bool CBlockTreeDB::LoadBlockIndexGuts(const Consensus::Params& consensusParams,
         if (pcursor->GetKey(key) && key.first == DB_BLOCK_INDEX) {
             CDiskBlockIndex diskindex;
             if (pcursor->GetValue(diskindex)) {
+                static_cast<CBlockIndex&>(diskindex).ToString();
                 // Construct block index object
                 CBlockIndex* pindexNew = insertBlockIndex(diskindex.GetBlockHash());
                 pindexNew->pprev          = insertBlockIndex(diskindex.hashPrev);
```

```
==563291== Invalid read of size 16
==563291==    at 0x2192B0: GetBlockHash (chain.h:266)
==563291==    by 0x2192B0: CBlockIndex::ToString[abi:cxx11]() const (chain.h:309)
==563291==    by 0x3CD954: CBlockTreeDB::LoadBlockIndexGuts(Consensus::Params const&, std::function<CBlockIndex* (uint256 const&)>) (txdb.cpp:312)
==563291==    by 0x210BBF: node::BlockManager::LoadBlockIndex(Consensus::Params const&) (blockstorage.cpp:258)
==563291==    by 0x2114AD: node::BlockManager::LoadBlockIndexDB(Consensus::Params const&) (blockstorage.cpp:323)
==563291==    by 0x410C0B: ChainstateManager::LoadBlockIndex() (validation.cpp:4148)
==563291==    by 0x222EEB: node::LoadChainstate(bool, ChainstateManager&, CTxMemPool*, bool, bool, long, long, long, bool, bool, std::function<bool ()>, std::function<void ()>) (chainstate.cpp:53)
==563291==    by 0x1749A4: AppInitMain(node::NodeContext&, interfaces::BlockAndHeaderTipInfo*) (init.cpp:1456)
==563291==    by 0x159A31: AppInit (bitcoind.cpp:234)
==563291==    by 0x159A31: main (bitcoind.cpp:278)
==563291==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==563291== 
==563291== 
==563291== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==563291==  Access not within mapped region at address 0x0
==563291==    at 0x2192B0: GetBlockHash (chain.h:266)
==563291==    by 0x2192B0: CBlockIndex::ToString[abi:cxx11]() const (chain.h:309)
==563291==    by 0x3CD954: CBlockTreeDB::LoadBlockIndexGuts(Consensus::Params const&, std::function<CBlockIndex* (uint256 const&)>) (txdb.cpp:312)
==563291==    by 0x210BBF: node::BlockManager::LoadBlockIndex(Consensus::Params const&) (blockstorage.cpp:258)
==563291==    by 0x2114AD: node::BlockManager::LoadBlockIndexDB(Consensus::Params const&) (blockstorage.cpp:323)
==563291==    by 0x410C0B: ChainstateManager::LoadBlockIndex() (validation.cpp:4148)
==563291==    by 0x222EEB: node::LoadChainstate(bool, ChainstateManager&, CTxMemPool*, bool, bool, long, long, long, bool, bool, std::function<bool ()>, std::function<void ()>) (chainstate.cpp:53)
==563291==    by 0x1749A4: AppInitMain(node::NodeContext&, interfaces::BlockAndHeaderTipInfo*) (init.cpp:1456)
==563291==    by 0x159A31: AppInit (bitcoind.cpp:234)
==563291==    by 0x159A31: main (bitcoind.cpp:278)
```

So I'd say the comment should be clarified that it is wholly illegal to call any such method on this object (whether through itself or a class it inherits from).

Reply:
Will update the comment and may add following belt-and-suspenders patch that fixes this case.

```diff
    std::string ToString() const
    {
        return strprintf("CBlockIndex(pprev=%p, nHeight=%d, merkle=%s, hashBlock=%s)",
            pprev, nHeight,
            hashMerkleRoot.ToString(),
-            GetBlockHash().ToString());
+            phashBlock ? GetBlockHash().ToString() : "");
    }
```

Reply:
or

```diff
     std::string ToString() const
     {
+        if (!phashBlock) return "";
         return strprintf("CBlockIndex(pprev=%p, nHeight=%d, merkle=%s, hashBlock=%s)",
```

Reply:
> I am pretty sure that constructing a CDiskBlockIndex, then calling the base class's method on the constructed object will result in dereferencing a nullptr (UB)

I see your point now.

```cpp
CDiskBlockIndex diskindex;
// Will crash because `phashBlock` is nullptr.
static_cast<CBlockIndex&>(diskindex).ToString();
```

But then, the following will also crash for the same reason:

```cpp
CBlockIndex index;
index.ToString();
```

The crash is not related to the inheritance or the deleted methods from this PR.

Reply:
> The crash is not related to the inheritance or the deleted methods from this PR.

Yes. I'm thinking it would maybe be best to revert this PR to the commit of the last ACK and propose the CBlockIndex improvements separately.

Reply:
> The crash is not related to the inheritance or the deleted methods from this PR.

Correct, this is why I commented on the documentation, which seemed to encourage calling the deleted method on the base class instead.

Reply:
Dropped the documentation and moved the fix that ensures phashBlock in CBlockIndex#ToString() is not nullptr before dereferencing to the first commit of this changeset. Don't hesitate to propose a code comment if you think one is still a good idea.